src/plugins/dist_matrixops/retile_annotations.cpp

```diff
@@ -0,0 +1,811 @@
+                )
+            )"},
+            &create_retile_annotations,
+            &execution_tree::create_primitive<retile_annotations>, R"(
+            a, tiling_type, intersection, numtiles, new_tiling
+            Args:
+
+                a (array): a distributed array. A vector or a matrix.
+                tiling_type (string, optional): defaults to `sym` which is a
+                    balanced way of tiling among all localities. Other options
+                    are `row`, `column` or `user` tiling types. If an
+                    intersection is given, retile produces tiles that have
+                    overlapped parts. In the `user` mode, using new_tiling,
+                    tiles are specified with their spans.
+                intersection (int or tuple of ints, optional): the size of
+                    overlapped part on each dimension. If an integer is given,
+                    that would be the intersection length on all dimensions
+                    that are tiled. If the given intersection is odd, the extra
+                    overlapped vector will be at the end of the part.
+                    In the `user` mode, `intersection` cannot be used.
+                numtiles (int, optional): number of tiles of the returned array
+                    If not given it sets to the number of localities in the
+                    application.
+                new_tiling (list, optional): a new tiling specification for the
+                    current locality that is specified only for the `user` mode
+                    E.g. for a matrix we can have:
+                    list("tile", list("columns", 0, 2), list("rows", 0, 2)) or
+                    list("args", list("locality", 0, 4),
+                       list("tile", list("columns", 0, 2), list("rows", 0, 2)))
+
+            Returns:
+
+            A retiled array according to the tiling type on numtiles localities
+            The returned array can contain overlapped parts)")
+    };
+
+    ///////////////////////////////////////////////////////////////////////////
+    retile_annotations::retile_annotations(
+        execution_tree::primitive_arguments_type&& operands,
+        std::string const& name, std::string const& codename)
+      : primitive_component_base(std::move(operands), name, codename)
+    {}
+
+    ///////////////////////////////////////////////////////////////////////////
+    namespace detail
+    {
+        std::tuple<bool, std::int64_t, std::int64_t> tile_extraction_1d_helper(
+            ir::range_iterator&& it)
+        {
+            ir::range tile_data = extract_list_value_strict(*++it);
+            ir::range_iterator inner_it = tile_data.begin();
+
+            // type_: columns = 0, rows = 1
+            bool type_ =
+                extract_string_value_strict(*inner_it) == "rows" ? 1 : 0;
+            std::int64_t start =
+                extract_scalar_nonneg_integer_value_strict(*++inner_it);
```

Comment:
All `extract_...` functions take the name and filename argument (they default to `"<unknown>"`). Error messages are much more readable if those are provided, however.